auto_round_extension/qbits/qlinear_qbits.py

```diff
@@ -0,0 +1,308 @@
+
+# Copyright (c) 2024 Intel Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import math
+
+import numpy as np
+import torch
+import torch.nn as nn
+import transformers
+from intel_extension_for_transformers import qbits  # with QBits kernels ()
+from auto_round.utils import convert_dtype_torch2str, logging
+
+logger = logging.getLogger(__name__)
+
+BITS_DTYPE_MAPPING = {
+    4: "int4_clip",
+    8: "int8",
+}
+
+
+class QuantLinear(nn.Module):
+    QUANT_TYPE = "qbits"
+
+    def __init__(
+        self,
+        bits,
+        group_size,
+        infeatures,
+        outfeatures,
+        bias,
+        kernel_switch_threshold=128,
+        trainable=False,
+        weight_dtype=torch.bfloat16,
+        **kwargs,
+    ):
+        super().__init__()
+
+        if bits not in [4, 8]:
+            raise NotImplementedError("Only 4,8 bits are supported for QBits.")
+
+        self.infeatures = infeatures
+        self.outfeatures = outfeatures
+        self.bits = bits
+        self.group_size = group_size if group_size != -1 else infeatures
+        self.maxq = 2**self.bits - 1
+        self.weight_dtype = weight_dtype
+        self.asym = True
+        self.need_repack = True
+
+        self.register_buffer(
+            "qweight",
+            torch.zeros((infeatures // 32 * self.bits,
+                        outfeatures), dtype=torch.int32),
+        )
+        self.register_buffer(
+            "qzeros",
+            torch.zeros(
+                (
+                    math.ceil(infeatures / self.group_size),
+                    outfeatures // 32 * self.bits,
+                ),
+                dtype=torch.int32,
+            ),
+        )
+        self.register_buffer(
+            "scales",
+            torch.zeros(
+                (math.ceil(infeatures / self.group_size), outfeatures),
+                dtype=weight_dtype,
+            ),
+        )
+        if bias:
+            self.register_buffer("bias", torch.zeros(
+                (outfeatures), dtype=torch.float))
+        else:
+            self.bias = None
+
+        self.kernel_switch_threshold = kernel_switch_threshold
+
+        self.trainable = trainable
+
+    def post_init(self):
+        assert self.qweight.device.type == "cpu"
```

Comment:
We still need to support models packed with ExLlama2, so I guess this function should be called in  auto_quantier.py

Reply:
done in 9cbc870